sled-agent/src/services.rs

```diff
@@ -779,17 +787,22 @@ impl ServiceManager {
 
         let running_zone = RunningZone::boot(installed_zone).await?;
 
-        for link in running_zone.links() {
-            info!(
-                self.inner.log,
-                "Ensuring {}/{} exists in zone",
-                link.name(),
-                IPV6_LINK_LOCAL_NAME
-            );
-            Zones::ensure_has_link_local_v6_address(
-                Some(running_zone.name()),
-                &AddrObject::new(link.name(), IPV6_LINK_LOCAL_NAME).unwrap(),
-            )?;
+        for (link, needs_link_local) in
+            running_zone.links().iter().zip(links_need_link_local)
```

Comment:
On a lark I tried using `.filter_map()` here (since we're already doing `.iter().zip(..)`):

```diff
@@ -787,22 +787,21 @@ impl ServiceManager {

         let running_zone = RunningZone::boot(installed_zone).await?;

-        for (link, needs_link_local) in
-            running_zone.links().iter().zip(links_need_link_local)
+        for link in
+            running_zone.links().iter().zip(links_need_link_local).filter_map(
+                |(link, needs_link_local)| needs_link_local.then_some(link),
+            )
         {
-            if needs_link_local {
-                info!(
-                    self.inner.log,
-                    "Ensuring {}/{} exists in zone",
-                    link.name(),
-                    IPV6_LINK_LOCAL_NAME
-                );
-                Zones::ensure_has_link_local_v6_address(
-                    Some(running_zone.name()),
-                    &AddrObject::new(link.name(), IPV6_LINK_LOCAL_NAME)
-                        .unwrap(),
-                )?;
-            }
+            info!(
+                self.inner.log,
+                "Ensuring {}/{} exists in zone",
+                link.name(),
+                IPV6_LINK_LOCAL_NAME
+            );
+            Zones::ensure_has_link_local_v6_address(
+                Some(running_zone.name()),
+                &AddrObject::new(link.name(), IPV6_LINK_LOCAL_NAME).unwrap(),
+            )?;
         }
```

I don't feel like either version is clearly better; take your pick.